### PR TITLE
LOG-1507: Consume EO's delete by query api changes

### PR DESCRIPTION
**Pull request by sasagarw**
### Description

This PR:
- extends clusterLogging API for delete by query functionality.
- updates go.mod file to use latest EO image.

/cc @shwetaap 
/assign @jcantrill 

### Links

- Depending on PR(s): https://github.com/openshift/origin-aggregated-logging/pull/2229
- JIRA: https://issues.redhat.com/browse/LOG-1507

### CODE REVIEW

**Comment by sasagarw** [comment]
/hold
This (https://github.com/openshift/elasticsearch-operator/pull/824) needs to be merged and then update the EO image to latest

File: internal/k8shandler/indexmanagement/index_management.go [hunk 1]
```diff
@@ -76,30 +76,59 @@ func newDefaultPoliciesSpec(spec *logging.RetentionPoliciesSpec) *logging.Retent
...
 		},
 		Infra: &logging.RetentionPolicySpec{
-			MaxAge: esapi.TimeUnit("7d"),
+			MaxAge:                  esapi.TimeUnit("7d"),
+			PruneNamespacesInterval: esapi.TimeUnit("30m"),
 		},
 		Audit: &logging.RetentionPolicySpec{
-			MaxAge: esapi.TimeUnit("7d"),
+			MaxAge:                  esapi.TimeUnit("7d"),
+			PruneNamespacesInterval: esapi.TimeUnit("30m"),
```
Comment (shwetaap) [inline internal/k8shandler/indexmanagement/index_management.go:88]: Are we trying to maintain the default interval as 15m or 30m? I thought you said 15m when we spoke.

Comment (sasagarw) [reply internal/k8shandler/indexmanagement/index_management.go:88]: I said 30m for delete by query thing. It was 15m for normal index management.

Comment (shwetaap) [reply internal/k8shandler/indexmanagement/index_management.go:88]: Ah ok! thanks

File: apis/logging/v1/clusterlogging_types.go [hunk 2]
```diff
@@ -145,6 +145,14 @@ type RetentionPoliciesSpec struct {
 type RetentionPolicySpec struct {
 	// +optional
 	MaxAge elasticsearch.TimeUnit `json:"maxAge"`
+
+	// How often to run a new prune-namespaces job
+	// +optional
+	PruneNamespacesInterval elasticsearch.TimeUnit `json:"pruneNamespacesInterval"`
+
+	// The per namespace specification to delete documents older than a given minimum age
```
Comment (vimalk78) [inline apis/logging/v1/clusterlogging_types.go:153]: should the comment say "a given minimum age" or a "a given maximum age" ?

File: config/crd/bases/logging.openshift.io_clusterloggings.yaml [hunk 3]
```diff
@@ -368,6 +368,25 @@ spec:
                             description: TimeUnit is a time unit like h,m,d
                             pattern: ^([0-9]+)([yMwdhHms]{0,1})$
                             type: string
+                          namespaceSpec:
+                            description: The per namespace specification to delete documents older than a given minimum age
+                            items:
+                              properties:
+                                minAge:
+                                  description: Delete the records matching the namespaces which are older than this MinAge (e.g. 1d)
```
Comment (vimalk78) [inline config/crd/bases/logging.openshift.io_clusterloggings.yaml:376]: is there a `maxAge` somewhere in ES CRD? if there is no min/max then just name it something like `agelimit`, or `maxAge`
If this is used to delete older namespaces, then calling it a minimum of some quantity sounds confusing.

File: config/crd/bases/logging.openshift.io_clusterloggings.yaml [hunk 4]
```diff
@@ -368,6 +368,25 @@ spec:
...
+                                  pattern: ^([0-9]+)([yMwdhHms]{0,1})$
+                                  type: string
+                                namespace:
```
Comment (vimalk78) [inline config/crd/bases/logging.openshift.io_clusterloggings.yaml:379]: if this is to be mostly used a prefix, name it as a prefix , like`nsPrefix`
usage should be clear from name. a prefix can be understood to be used as exact name, but exact name may not be interpreted to be used as prefix.

File: internal/k8shandler/indexmanagement/index_management.go [hunk 5]
```diff
@@ -113,10 +142,16 @@ func newPolicySpec(name string, maxIndexAge esapi.TimeUnit, hotPhaseAge esapi.Ti
 				},
 			},
 			Delete: &esapi.IndexManagementDeletePhaseSpec{
-				MinAge: maxIndexAge,
+				MinAge:                  retentionPolicy.MaxAge,
```
Comment (vimalk78) [inline internal/k8shandler/indexmanagement/index_management.go:145]: MinAge assigned from MaxAge. (deleted line also doing same though)

**Review by vimalk78 — COMMENTED** [review]
some comments

**Comment by sasagarw** [comment]
@vimalk78 I understand your naming concern in API which all your comment tells. I have kept same name as the spike document was finalized. Also, EO has already implemented this feature after much discussions and if I change the names in CLO then there would be inconsistency. Check the EO PR https://github.com/openshift/elasticsearch-operator/pull/797

File: internal/k8shandler/indexmanagement/index_management.go (hunk 5 above)
Comment (sasagarw) [reply internal/k8shandler/indexmanagement/index_management.go:145]: Deleted line was doing same but I wanted more variables which is part of this `retentionPolicy` variable. I made the function definition generic and removed the individual variable.

**Comment by vimalk78** [comment]
> @vimalk78 I understand your naming concern in API which all your comment tells. I have kept same name as the spike document was finalized. Also, EO has already implemented this feature after much discussions and if I change the names in CLO then there would be inconsistency. Check the EO PR [openshift/elasticsearch-operator#797](https://github.com/openshift/elasticsearch-operator/pull/797)

Yes, these comments are mostly on EO CRD

**Pull request merged**